OpenNMT/train.py

```diff
@@ -0,0 +1,280 @@
+                    help="""For this many epochs, order the minibatches based
+                    on source sequence length. Sometimes setting this to 1 will
+                    increase convergence speed.""")
+parser.add_argument('-pre_word_vecs_enc',
+                    help="""If a valid path is specified, then this will load
+                    pretrained word embeddings on the encoder side.
+                    See README for specific formatting instructions.""")
+parser.add_argument('-pre_word_vecs_dec',
+                    help="""If a valid path is specified, then this will load
+                    pretrained word embeddings on the decoder side.
+                    See README for specific formatting instructions.""")
+
+# GPU
+parser.add_argument('-cuda', action='store_true',
+                    help="Use CUDA")
+
+parser.add_argument('-log_interval', type=int, default=50,
+                    help="Print stats at this interval.")
+# parser.add_argument('-seed', type=int, default=3435,
+#                     help="Seed for random initialization")
+
+opt = parser.parse_args()
+print(opt)
+
+if torch.cuda.is_available() and not opt.cuda:
+    print("WARNING: You have a CUDA device, so you should probably run with -cuda")
+
+
+def NMTCriterion(vocabSize):
+    weight = torch.ones(vocabSize)
+    weight[onmt.Constants.PAD] = 0
+    crit = nn.NLLLoss(weight, size_average=False)
+    if opt.cuda:
+        crit.cuda()
+    return crit
+
+
+def memoryEfficientLoss(outputs, targets, generator, crit, eval=False):
+    # compute generations one piece at a time
+    loss = 0
+    outputs = Variable(outputs.data, requires_grad=(not eval), volatile=eval)
+
+    batch_size = outputs.size(1)
+    outputs_split = torch.split(outputs, opt.max_generator_batches)
+    targets_split = torch.split(targets, opt.max_generator_batches)
+    for out_t, targ_t in zip(outputs_split, targets_split):
+        out_t = out_t.view(-1, out_t.size(2))
+        pred_t = generator(out_t)
+        loss_t = crit(pred_t, targ_t.view(-1))
+        loss += loss_t.data[0]
+        if not eval:
+            loss_t.div(batch_size).backward()
+
+    grad_output = None if outputs.grad is None else outputs.grad.data
+    return loss, grad_output
+
+
+def eval(model, criterion, data):
+    total_loss = 0
+    total_words = 0
+
+    model.eval()
+    for i in range(len(data)):
+        batch = data[i]
+        outputs = model(batch)  # FIXME volatile
+        targets = batch[1][1:]  # exclude <s> from targets
+        loss, _ = memoryEfficientLoss(
+                outputs, targets, model.generator, criterion, eval=True)
+        total_loss += loss
+        total_words += targets.data.ne(onmt.Constants.PAD).sum()
+
+    model.train()
+    return total_loss / total_words
+
+
+def trainModel(model, trainData, validData, dataset, optim):
+    print(model)
+    model.train()
+    for p in model.parameters():
+        p.data.uniform_(-opt.param_init, opt.param_init)
+
```

Comment:
I think the two lines above will reinitialize parameters even if the model is loaded from a checkpoint. Potential fix at https://github.com/adamlerer/examples/pull/2